src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -79,9 +81,101 @@ protected override void AddInterfaceImplementationCore([DynamicallyAccessedMembe
 
         [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2083:DynamicallyAccessedMembers", Justification = "Not sure how to handle")]
         [return: DynamicallyAccessedMembers((DynamicallyAccessedMemberTypes)(-1))]
-        protected override TypeInfo CreateTypeInfoCore() => this;
-        protected override ConstructorBuilder DefineConstructorCore(MethodAttributes attributes, CallingConventions callingConvention, Type[]? parameterTypes, Type[][]? requiredCustomModifiers, Type[][]? optionalCustomModifiers) => throw new NotImplementedException();
-        protected override ConstructorBuilder DefineDefaultConstructorCore(MethodAttributes attributes) => throw new NotImplementedException();
+        protected override TypeInfo CreateTypeInfoCore()
+        {
+            if (_isCreated)
+            {
+                return this;
+            }
+
+            // Create a public default constructor if this class has no constructor. Except the type is Interface, ValueType, Enum, or a static class.
+            // (TypeAttributes.Abstract | TypeAttributes.Sealed) determines if the type is static
+            if (_constructorDefinitions.Count == 0 && (_attributes & TypeAttributes.Interface) == 0 && !IsValueType &&
+                ((_attributes & (TypeAttributes.Abstract | TypeAttributes.Sealed)) != (TypeAttributes.Abstract | TypeAttributes.Sealed)))
+            {
+                DefineDefaultConstructor(MethodAttributes.Public);
+            }
+
```

Comment:
Do we also need to populate method and field tokens here? (It can wait for follow up PR.)

The runtime and .NET Framework impl let you access the tokens of baked fields and methods.

Reply:
Hm, the tokens could not be populated reliably until save or all types are baked.

Reply:
Why not?

Reply:
It is about def tokens for members of the baked type only.

Reply:
Yah, I guess we can force the baked types order to make it work

Reply:
Done, now types need to be baked before save